src/charm.py

```diff
@@ -0,0 +1,197 @@
+    import bcrypt
+
+
+class Operator(CharmBase):
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+        self.log = logging.getLogger(__name__)
+        self.image = OCIImageResource(self, "oci-image")
+        self.service_mesh = ServiceMeshRequires(self, "service-mesh")
+
+        self._stored.set_default(username="admin")
+        self._stored.set_default(
+            password="".join(random.choices(string.ascii_letters, k=30))
+        )
+        self._stored.set_default(salt=bcrypt.gensalt)
+        self._stored.set_default(user_id=str(uuid4()))
+
+        for event in [
+            self.on.install,
+            self.on.leader_elected,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on.oidc_client_relation_changed,
+        ]:
+            self.framework.observe(event, self.main)
+
+        self.framework.observe(
+            self.on.service_mesh_relation_joined, self.configure_mesh
+        )
+        self.framework.observe(
+            self.on.service_mesh_relation_changed, self.configure_mesh
+        )
+
+    def configure_mesh(self, event):
+        self.log.info("Adding route /dex to service mesh")
+        self.service_mesh.add_route(
+            prefix="/dex",
+            service=self.model.app.name,
+            port=self.model.config["port"],
+        )
+
+    def main(self, event):
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            self.log.info(e)
+            return
+
+        self.model.unit.status = MaintenanceStatus("Setting pod spec")
+
+        connectors = yaml.safe_load(self.model.config["connectors"])
+        port = self.model.config["port"]
+        public_url = self.model.config["public-url"]
+
+        oidc_client_info = []
+        if "oidc-client" in self.model.relations:
+            for relation in self.model.relations["oidc-client"]:
+                for unit in relation.units:
+                    if unit in relation.data:
+                        relation_data = dict(relation.data[unit])
+                        oidc_client_info.append(relation_data)
+                        break
+
+        # Allows setting a basic username/password combo
+        static_username = self.model.config["static-username"]
+        static_password = self.model.config["static-password"]
+
+        static_config = {}
+
+        # Dex needs some way of logging in, so if nothing has been configured,
+        # just generate a username/password
+        if not static_username:
+            static_username = self._stored.username
+
+        if static_username:
```

Comment:
This logic was copied over from the existing charm, but in re-reading it, I think this is essentially `if True:` due to the above `if not static_username` check, and can be skipped. I originally wanted to allow someone to disable the static username/password generation, but I don't think that's necessary, since even if they're using LDAP or some other scheme, they can just have an unused static username/password combo.